Ambient L4/ ZTunnel User's Guide   Issue #13481

srampal opened the pull request:
Please provide a description for what this PR is for.

This PR is for the issue  https://github.com/istio/istio.io/issues/13481 and will include commits for the initial portion of the Ambient Usage Guide. 

And to help us figure out who should review this PR, please 
put an X in all the areas that this PR affects.

- [X] Ambient
- [X] Docs
- [] Installation
- [] Networking
- [ ] Performance and Scalability
- [ ] Extensions and Telemetry
- [ ] Security
- [ ] Test and Release
- [] User Experience
- [ ] Developer Infrastructure

istio-testing commented:
Hi @srampal. Thanks for your PR.

I'm waiting for a [istio](https://github.com/orgs/istio/people) member to verify that this patch is reasonable to test. If it is, they should reply with `/ok-to-test` on its own line. Until that is done, I will not automatically test new commits in this PR, but the usual testing commands by org members will still work. Regular contributors should [join the org](https://github.com/orgs/istio/people) to skip this step.

Once the patch is verified, the new status will be reflected by the `ok-to-test` label.

I understand the commands that are listed [here](https://go.k8s.io/bot-commands?repo=istio%2Fistio.io).

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository.
</details>

srampal commented:
@howardjohn @linsun please comment. Also it is still unclear how we want to structure the Ambient usage guide so the current PR is just to create an initial skeleton, divide the guide into multiple sub-guides and start with Ztunnel & L4 networking as the first  sub-guide. Would like to review further with you.

ericvn commented:
/ok-to-test

istio-testing commented:
@srampal: The following test **failed**, say `/retest` to rerun all failed tests or `/retest-required` to rerun all mandatory failed tests:

Test name | Commit | Details | Required | Rerun command
--- | --- | --- | --- | ---
lint_istio.io | b46c4d9afe5633714915d6972443ceb5e7727c05 | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio.io/13635/lint_istio.io/1684192757893042176) | true | `/test lint`

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

istio-testing commented:
@srampal: The following test **failed**, say `/retest` to rerun all failed tests or `/retest-required` to rerun all mandatory failed tests:

Test name | Commit | Details | Required | Rerun command
--- | --- | --- | --- | ---
lint_istio.io | 52691df4b935685ef9191cd8692bfaf2427d40e0 | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio.io/13635/lint_istio.io/1690095283314102272) | true | `/test lint`

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

istio-testing commented:
@srampal: The following test **failed**, say `/retest` to rerun all failed tests or `/retest-required` to rerun all mandatory failed tests:

Test name | Commit | Details | Required | Rerun command
--- | --- | --- | --- | ---
lint_istio.io | c7e1373b2ed4558d16538bd311b784f6c2fe04cc | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio.io/13635/lint_istio.io/1694076479379345408) | true | `/test lint`

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

istio-testing commented:
@srampal: The following test **failed**, say `/retest` to rerun all failed tests or `/retest-required` to rerun all mandatory failed tests:

Test name | Commit | Details | Required | Rerun command
--- | --- | --- | --- | ---
lint_istio.io | 3df7e2d5f4275bbdc76aaea5398827ae69e64eac | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio.io/13635/lint_istio.io/1703895803178782720) | true | `/test lint`

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

srampal commented:
@howardjohn @linsun @bleggett this is still work in progress but inviting your comments to see if we agree on the content so far and content organization/ layout direction ...

bleggett commented on content/en/docs/ops/ambient/usage/ztunnel/_index.md:
```diff
@@ -0,0 +1,498 @@
+These log messages confirm the traffic indeed used the `ztunnel` proxy and traffic redirection between pods and the proxy is likely working correctly. Additional fine grained monitoring can be done by checking logs on the specific `ztunnel` proxy instances that are on the same nodes as the source and destination pods of traffic.  If these logs are not seen, then a possibility is that traffic redi ...
+
+### Checking for ztunnel traffic redirection
+In case the traffic logs for the `ztunnel` proxy do not indicate forwarding of traffic, you can check whether traffic redirection has been set up correctly. In this section we do not go into the full details of how traffic redirection is performed, but we provide some quick pointers to check.
```
I'm of a mind that all the `iptables` and `ipset` stuff here should be moved out into an "advanced debugging" guide, that lives in the devwiki/dev README - it's tightly tied to the current impl, and it's far more than we need to ask users in order to get a good bug report, and far more than what users need to validate that ambient is working in their cluster.

The goal of the docs here should be equipping people to know where to check that that ztunnel is capturing traffic, and to know where to go to collect the correct logs (`ztunnel` and `istio-cni`) if something isn't working.

srampal replied:
okay, will remove the subsections related to traffic redirection and add a note that this content will be in a separate guide.

bleggett commented on content/en/docs/ops/ambient/usage/ztunnel/_index.md:
```diff
@@ -0,0 +1,498 @@
+
+<< To add some content on the xDS api objects of relevance to Ambient/ ztunnel >>
+
+## Deploying an Application
```
```suggestion
## Deploying an Application
```

I would suggest breaking this up into 2 separate back-to-back sections for better clarity and impact:

`## Deploying Applications`

`## Adding Deployed Applications To The Ambient Mesh`

srampal replied:
Okay, making this change.

bleggett commented on content/en/docs/ops/ambient/usage/ztunnel/_index.md:
```diff
@@ -0,0 +1,498 @@
+
+TODO any other checks/ monitoring to be added ?
+
+### Pod selection logic for Ambient and Sidecar modes
+Istio with sidecar proxies can co-exist with `ambient` based node level proxies within the same compute cluster. Since there are multiple configuration parameters and resource annotations that govern whether a given pod will be setup to use a sidecar proxy or an ambient node proxy, the logic to make this decision is as follows.
+
+1. The `istio-cni` plugin configuration exclude list configured in `cni.values.excludeNamespaces` is used to skipnamespaces in the exclude list.
+2. `ambient` mode is used for a pod if 
+- The namespace has label "istio.io/dataplane-mode" == "ambient" 
+- The annotation "sidecar.istio.io/status" is not present on the pod
+- "ambient.istio.io/redirection" is not "disabled"
+
```
Suggest we move this section up so it comes right after the `#Adding workloads to ambient mesh` section, or maybe include this piece in that section.

srampal replied:
Retaining the order of this section for now but added a pointer to it in the Enabling Ambient section. Since we want the first pass of users to work with simple pure `ambient` only scenarios, it would be better to complete both deploymewnt and monitoring of pure ambient before getting into mixed scenarios of ambient + sidecar.j

srampal commented on content/en/docs/ops/ambient/usage/ztunnel/_index.md:
```diff
@@ -0,0 +1,498 @@
+
+This guide describes the functionality and usage of the Ztunnel proxy and basic Layer-4 networking functions (also called Secure Overlay Networking) using Istio Ambient mesh. We use sample user journeys to describe these functions. 
+
+The Ztunnel (Zero Trust Tunnel) component is a purpose-built per-node proxy for Istio ambient mesh. It is responsible for securely connecting and authenticating workloads within ambient mesh. Ztunnel is designed to focus on a small set of features for your workloads in ambient mesh such as mTLS, authentication, L4 authorization and telemetry, without terminating workload HTTP traffic or parsing w ...
```
Agree. Am making the suggested changes with a slight tweak to the list line as follows "The ztunnel ensures L3 and L4 traffic is efficiently and securely transported to the waypoint proxies, where the full suite of Istio’s L7 functionality, such as HTTP telemetry and load balancing, is implemented."

srampal commented on content/en/docs/ops/ambient/usage/ztunnel/_index.md:
```diff
@@ -0,0 +1,498 @@
+
+The Ztunnel (Zero Trust Tunnel) component is a purpose-built per-node proxy for Istio ambient mesh. It is responsible for securely connecting and authenticating workloads within ambient mesh. Ztunnel is designed to focus on a small set of features for your workloads in ambient mesh such as mTLS, authentication, L4 authorization and telemetry, without terminating workload HTTP traffic or parsing w ...
+
+It is expected that some production use cases of Istio in Ambient mode may be addressed solely via the L4 Secure overlay networking features whereas other use cases will additionally need advanced traffic management and L7 networking features for which the additional Waypoint proxies will need to be deployed. This is summarized in the following table. This guide focusses on functionality related  ...
```
Agreed. Will make the suggested changes.

srampal commented on content/en/docs/ops/ambient/usage/ztunnel/_index.md:
```diff
@@ -0,0 +1,498 @@
+
+<< Consider breaking this out into bullets for easier reading TODO >>
+
+Ztunnel proxies are automatically installed when one of the supported installation methods is used to install Istio Ambient mesh.  The minimum Istio version required for the functionality described in this guide is 1.18.0. At this time, the ambient mode is only supported for deployment on Kubernetes clusters, support for deployment on non-Kubernetes endpoints such as Virtual machines is expected  ...
```
Agreed. Making the suggested changes.

srampal commented on content/en/docs/ops/ambient/usage/ztunnel/_index.md:
```diff
@@ -0,0 +1,498 @@
+
+<< Consider breaking this out into bullets for easier reading TODO >>
+
+One of the goals for the ztunnel proxy design is to provide a usable configuration out of the box with a fixed feature set and that does not require much, or any, custom configuration. Hence currently there are no configuration options that need to be set other than the `ambient` profile setting. Once this profile is used, this in turn sets sets 2 internal configuration parameters (as illustrated ...
```
Agreed, removing this chunk.

srampal commented on content/en/docs/ops/ambient/usage/ztunnel/_index.md:
```diff
@@ -0,0 +1,498 @@
+
+Normally, a user with Istio admin privileges will deploy the Istio mesh infrastructure. Once Istio is successfully deployed in `ambient` mode, it will be transparently available to applications deployed by all users in namespaces that have been annoted to use Istio `ambient` as illustrated in the examples below.  
+
+Lets first deploy a simple http client server application without making it part of the Istio ambient mesh. We can pick from the apps in the samples folder of the istio repository. Execute the following examples from the top of a local Istio repository or istio folder created by downloading the istioctl client as described in istio guides.
```
Agreed. Making this change.

srampal commented on content/en/docs/ops/ambient/usage/ztunnel/_index.md:
```diff
@@ -0,0 +1,498 @@
+    <title>httpbin.org</title>
+{{< /text >}}
+
+In this case, the traffic traversed an enecyrpted ztunnel. In the next section we look at how to check that traffic is using the ztunnel and some monitoring options. 
```
N/A now since this chunk was removed

AndreaM12345 commented on content/en/docs/ops/ambient/usage/ztunnel/index.md:
```diff
@@ -0,0 +1,545 @@
+
+
+[//]: # (There are additional subtle details on how functionality is split between the ztunnel and the waypoint proxies. These are described in later sections of this guide and other guides.)
+[//]: # ( As an example, if an Istio Authorization policy includes strictly L4 rules, then it is handled completely by ztunnel but if a policy includes only L7 rules or both L4 and L7 rules then it is handled by a Waypoint proxy. However when it comes to a VirtualService it is always handled by Waypoint proxies regardless of whether it has purely L4 rules, purely L7 rules or a mix of L4 and L7 ru ...
```
not clear about if auth policy has both L4 and L7 if its only handled by waypoint proxy...  I think its still under debate what should happen based on this issue https://github.com/istio/istio/issues/47137.  Not sure if we want to include this section until its more clear.  I do agree its useful knowledge just maybe needs to wait.

linsun replied:
Authz is going through a lot of change in the community, might be a little too early to write.

srampal replied:
This text is commented out. We will have a small example of L4 authz similar to the getting started guide for now as discussed on the ambient slack channel. Additional enhancements will be documented once we approach close to beta status.

AndreaM12345 commented on content/en/docs/ops/ambient/usage/ztunnel/index.md:
```diff
@@ -0,0 +1,545 @@
+
+<< Consider breaking this out into bullets for easier reading TODO >>
+
+One of the goals for the ztunnel proxy design is to provide a usable configuration out of the box with a fixed feature set and that does not require much, or any, custom configuration. Hence currently there are no configuration options that need to be set other than the `ambient` profile setting. Once this profile is used, this in turn sets sets 2 internal configuration parameters (as illustrated ...
```
Once this profile is used, this in turn sets sets 2 internal configuration parameters (as illustrated in the examples below) within the istioOperator which eventually sets the configuration of the `ambient` mesh. In the future there may be some additional limited configurability for ztunnel proxies.

srampal replied:
this section has been removed from this guide. There will be a separate installation sub-guide which is being covered by a separate PR.

bleggett commented on content/en/docs/ops/ambient/usage/ztunnel/ztunnel-datapath-1.png:
As mentioned - the way we talk about it now, `ztunnel` is the name of the proxy, the actual tunneled connection is an "HBONE tunnel" - or just a "tunnel" if we want to minimize occurrences of "HBONE" in the docs - so the tube in the middle probably needs a clearer name.

linsun replied:
agree ^^

srampal replied:
done. changed the text and the figure

linsun commented:
@srampal Thanks for the work - i think the guide is coming together nicely.

One feedback - I noticed there are quite a lot of places talking about in the future - i would prefer us not mention future in the docs as they could change and may not be accurate soon. I would prefer the guide to focus on the architecture, what works and what user can do with it.

cc @howardjohn @fykaa

linsun commented on content/en/docs/ops/ambient/usage/ztunnel/index.md:
```diff
@@ -0,0 +1,545 @@
+
+The Ztunnel (Zero Trust Tunnel) component is a purpose-built per-node proxy for Istio ambient mesh. It is responsible for securely connecting and authenticating workloads within ambient mesh. Ztunnel is designed to focus on a small set of features for your workloads in ambient mesh such as mTLS, authentication, L4 authorization and telemetry, without terminating workload HTTP traffic or parsing w ...
+
+It is expected that some production use cases of Istio in Ambient mode may be addressed solely via the L4 Secure overlay networking features whereas other use cases will additionally need advanced traffic management and L7 networking features for which the additional Waypoint proxies will need to be deployed. This is summarized in the following table. This guide focusses on functionality related  ...
```
would it be possible not mention L7 other than a simple link to keep this focus on L4?  The reason is user may not be interested in L7 at all.

srampal replied:
Yes, the guide is for L4. L7 is only mentioned where absolutely necessary to explain the L4 function or how ztunnel separates out the L4 from the L7. Same logic was used in other ztunnel L4 blog posts also.

linsun commented on content/en/docs/ops/ambient/usage/ztunnel/index.md:
```diff
@@ -0,0 +1,545 @@
+
+
+| Application Deployment Use Case | Istio Ambient Mesh Configuration |
+| ------------- | ------------- |
+| Zero Trust networking via mutual-TLS, encrypted and tunneled data transport of all IP protocols, L4 authorization,  L4 telemetry | Baseline Ambient Mesh with ztunnel proxy networking |
+| Application requires L4 Mutual-TLS plus advanced Istio traffic management features (incl L4/ L7 VirtualService, L7 telemetry, L7 Authorization) | Full Istio Ambient Mesh configuration both ztunnel proxy and waypoint proxy based networking |
```
feels this belongs more in the L7 guide.

cc @fykaa

srampal commented on content/en/docs/ops/ambient/usage/ztunnel/_index.md:
```diff
@@ -0,0 +1,498 @@
+These log messages confirm the traffic indeed used the `ztunnel` proxy and traffic redirection between pods and the proxy is likely working correctly. Additional fine grained monitoring can be done by checking logs on the specific `ztunnel` proxy instances that are on the same nodes as the source and destination pods of traffic.  If these logs are not seen, then a possibility is that traffic redi ...
+
+### Checking for ztunnel traffic redirection
+In case the traffic logs for the `ztunnel` proxy do not indicate forwarding of traffic, you can check whether traffic redirection has been set up correctly. In this section we do not go into the full details of how traffic redirection is performed, but we provide some quick pointers to check.
```
Okay, adding some content for checking istio-cni logs

srampal commented on content/en/docs/ops/ambient/usage/ztunnel/_index.md:
```diff
@@ -0,0 +1,498 @@
+description: User guide for ztunnel networking.
+weight: 2
+owner: istio/wg-networking-maintainers
+test: n/a
```
The code snippets have been tested on a live cluster. Since the current guide is alpha only and we want to get it out to assist early testers of ambient it would be better to get the first pass of this guide merged without adding the tests. Also some of the snippets involve trimmed output and may change in the second pass of updates before ambient goes to beta.  We can add the tests on a followup update to this guide as it approaches beta status. This was okay with the other reviewers, if needed we can discuss further, else for the first pass, we would like to merge wirthout the tests of the code snippets.

srampal commented on content/en/docs/ops/ambient/usage/ztunnel/_index.md:
```diff
@@ -0,0 +1,498 @@
+
+## Introduction 
+
+This guide describes the functionality and usage of the Ztunnel proxy and basic Layer-4 networking functions (also called Secure Overlay Networking) using Istio Ambient mesh. We use sample user journeys to describe these functions. 
```
okay

srampal commented on content/en/docs/ops/ambient/usage/ztunnel/_index.md:
```diff
@@ -0,0 +1,498 @@
+Ztunnel proxies are automatically installed when one of the supported installation methods is used to install Istio Ambient mesh.  The minimum Istio version required for the functionality described in this guide is 1.18.0. At this time, the ambient mode is only supported for deployment on Kubernetes clusters, support for deployment on non-Kubernetes endpoints such as Virtual machines is expected  ...
+
+{{< tip >}}
+A single Istio mesh can include pods and endpoints some of which operate using the sidecar proxy mode while others use the node level proxy of the Ambient architecture. 
```
okay

srampal commented on content/en/docs/ops/ambient/usage/ztunnel/_index.md:
```diff
@@ -0,0 +1,498 @@
+A single Istio mesh can include pods and endpoints some of which operate using the sidecar proxy mode while others use the node level proxy of the Ambient architecture. 
+{{< /tip >}}
+
+### Understanding the Ztunnel Default Configuration
```
okay

srampal commented on content/en/docs/ops/ambient/usage/ztunnel/_index.md:
```diff
@@ -0,0 +1,498 @@
+
+One of the goals for the ztunnel proxy design is to provide a usable configuration out of the box with a fixed feature set and that does not require much, or any, custom configuration. Hence currently there are no configuration options that need to be set other than the `ambient` profile setting. Once this profile is used, this in turn sets sets 2 internal configuration parameters (as illustrated ...
+
+For the examples in this guide, we used a deployment of Istio Ambient on a `kind` cluster, although these should apply for any Kubernetes cluster version 1.18.0 or later. Refer to the Getting started guide on how to download the `istioctl` client and how to deploy a `kind` cluster. It would be recommended to have a cluster with more than 1 worker node in order to fully exercise the examples descr ...
```
okay

srampal commented on content/en/docs/ops/ambient/usage/ztunnel/_index.md:
```diff
@@ -0,0 +1,498 @@
+
+### Installation using istioctl
+
+Setting Istio profile to `ambient` during installation is all that is needed to enable installation of Ztunnel and Layer-4 networking functionality. An instance of Istio mesh cannot be dynamically switched between sidecar mode and `ambient` mode. A prior instance of istio must be uninstalled before re-innstalling istio in `ambient` for the same set of user application endpoints and namespaces. 
```
okay

srampal commented on content/en/docs/ops/ambient/usage/ztunnel/_index.md:
```diff
@@ -0,0 +1,498 @@
+{{< /text >}}
+
+After running the above command, you’ll get the following output that indicates
+five components (including {{< gloss "ztunnel" >}}Ztunnel{{< /gloss >}}) have been installed successfully!
```
okay

srampal commented on content/en/docs/ops/ambient/usage/ztunnel/_index.md:
```diff
@@ -0,0 +1,498 @@
+{{< /text >}}
+
+After running the above command, you’ll get the following output that indicates
+four components (including {{< gloss "ztunnel" >}}Ztunnel{{< /gloss >}}) have been installed successfully!
```
okay

srampal commented on content/en/docs/ops/ambient/usage/ztunnel/_index.md:
```diff
@@ -0,0 +1,498 @@
+
+For the examples in this guide, we used a deployment of Istio Ambient on a `kind` cluster, although these should apply for any Kubernetes cluster version 1.18.0 or later. Refer to the Getting started guide on how to download the `istioctl` client and how to deploy a `kind` cluster. It would be recommended to have a cluster with more than 1 worker node in order to fully exercise the examples descr ...
+
+### Installation using istioctl
```
The entire installation section has been removed from this guide and will be moved to a separate sub-guide.

srampal commented:
> @srampal Thanks for the work - i think the guide is coming together nicely.
> 
> One feedback - I noticed there are quite a lot of places talking about in the future - i would prefer us not mention future in the docs as they could change and may not be accurate soon. I would prefer the guide to focus on the architecture, what works and what user can do with it.
> 
> cc @howardjohn @fykaa

@linsun sure .. I removed 4 of the future references and only retained 1 or 2 as discussed on the slack thread.

srampal commented:
@bleggett @dhawton @linsun @AndreaM12345 I have made further updates and also addressed your review comments. The guide is almost ready except for the very last section on metrics which I am not sure whether is needed at all. Other than that I think the guide is ready for your final review comments if any and then a merge. Note that we will still have a final pass of document updates right before the Beta release in order since some things are still in flux. However since the goal is to have some reasonably accurate docs for current ambient testers, we should probably get this merged with the understanding that the final documentation updates can be made just prior to the beta release date based on final release content.  Let me know if you have further comments/ questions. Thanks for your help and time spent answering my questions and reviewing this PR.

istio-testing commented:
@srampal: The following test **failed**, say `/retest` to rerun all failed tests or `/retest-required` to rerun all mandatory failed tests:

Test name | Commit | Details | Required | Rerun command
--- | --- | --- | --- | ---
lint_istio.io | 8d691c210d48823a48f4fd3e71c62efad0c53d7e | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio.io/13635/lint_istio.io/1717583971849605120) | true | `/test lint`

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

istio-testing commented:
@srampal: The following test **failed**, say `/retest` to rerun all failed tests or `/retest-required` to rerun all mandatory failed tests:

Test name | Commit | Details | Required | Rerun command
--- | --- | --- | --- | ---
lint_istio.io | 77d471be2e828df52fe7eff21cee7d3ba7d95a3e | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio.io/13635/lint_istio.io/1717592090512199680) | true | `/test lint`

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

istio-testing commented:
@srampal: The following test **failed**, say `/retest` to rerun all failed tests or `/retest-required` to rerun all mandatory failed tests:

Test name | Commit | Details | Required | Rerun command
--- | --- | --- | --- | ---
lint_istio.io | ee2eda9de1ad09c46635b067f1f51d342c58f3a7 | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio.io/13635/lint_istio.io/1717615947394584576) | true | `/test lint`

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

istio-testing commented:
@srampal: The following tests **failed**, say `/retest` to rerun all failed tests or `/retest-required` to rerun all mandatory failed tests:

Test name | Commit | Details | Required | Rerun command
--- | --- | --- | --- | ---
lint_istio.io | ee2eda9de1ad09c46635b067f1f51d342c58f3a7 | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio.io/13635/lint_istio.io/1717615947394584576) | true | `/test lint`
gencheck_istio.io | 5ddf1e08303ad54aa67da60a9dbe31fb29dac061 | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio.io/13635/gencheck_istio.io/1717637263078199296) | true | `/test gencheck`

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

istio-testing commented:
@srampal: The following tests **failed**, say `/retest` to rerun all failed tests or `/retest-required` to rerun all mandatory failed tests:

Test name | Commit | Details | Required | Rerun command
--- | --- | --- | --- | ---
gencheck_istio.io | 5ddf1e08303ad54aa67da60a9dbe31fb29dac061 | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio.io/13635/gencheck_istio.io/1717637263078199296) | true | `/test gencheck`
lint_istio.io | 5ddf1e08303ad54aa67da60a9dbe31fb29dac061 | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio.io/13635/lint_istio.io/1717637263023673344) | true | `/test lint`

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

istio-testing commented:
@srampal: The following test **failed**, say `/retest` to rerun all failed tests or `/retest-required` to rerun all mandatory failed tests:

Test name | Commit | Details | Required | Rerun command
--- | --- | --- | --- | ---
lint_istio.io | 2b2f8d0db3c143d84789653737ae76e93c8fb8ef | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio.io/13635/lint_istio.io/1717642630952652800) | true | `/test lint`

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

istio-testing commented:
@srampal: The following test **failed**, say `/retest` to rerun all failed tests or `/retest-required` to rerun all mandatory failed tests:

Test name | Commit | Details | Required | Rerun command
--- | --- | --- | --- | ---
lint_istio.io | 0158d890ed67c31bb094a48f9c2bbc501096e76e | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio.io/13635/lint_istio.io/1717653957377003520) | true | `/test lint`

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

ilrudie commented on content/en/docs/ops/ambient/usage/ztunnel/index.md:
```diff
@@ -0,0 +1,656 @@
+We note that the figure also shows traffic from pod C3 destined to destination pod S1 on worker node W2 also traverses the local ztunnel proxy instance so that L4 traffic management functions such as L4 Authorization and L4 Telemetry are enforced identically on traffic whether or not it crosses a node boundary. 
+
+#### Ztunnel datapath via waypoint
+The next figure illustrates the data path for a different use case where there is a need to traverse through an interim Waypoint proxy. This is the case when the application service requires advanced L7 traffic routing, management or policy handling. Here ztunnel L4 networking is used to tunnel traffic torwads a Waypoint proxy for L7 processing and then mapped via a second HBONE tunnel towards th ...
```
```suggestion
The next figure depicts the data path for a use case which requires advanced L7 traffic routing, management or policy handling. Here ztunnel uses HBONE tunneling to send traffic to a waypoint proxy for L7 processing. After processing, the waypoint sends traffic via a second HBONE tunnel to the ztunnel on the node hosting the destination pod.  In general the waypoint proxy may or may not be located on the same nodes as the source or destination pod. 
```

srampal replied:
okay .. done.

ilrudie commented on content/en/docs/ops/ambient/usage/ztunnel/index.md:
```diff
@@ -0,0 +1,656 @@
+    >}}
+
+
+#### Ztunnel datapath hairpinning 
```
It may be better to leave this section out. It seems like the details and plans of hair pinning are too fluid for this type of document. It's an interesting area with lots of discussion but at least to my mind it's just not really pinned down enough yet.

srampal replied:
This is the current behavior. Since this guide is meant to help alpha testers who will encounter this functionality in Istio 1.18, 1.19 and possibly permanently, I think it might be better to leave this in the guide until such a time that the community does decide to change it. To be clear, I am adding an additional statement that this behavior is currently under review and could change in future.

srampal replied:
Added a warning blurb as follows .. to make this more explicit for hair-pinning even though there is already a disclaimer at the top of the guide about any of the functionality potentially changing as the project evolves to beta and beyond.
_{{< warning >}}
As noted earlier, some ambient functions may change as the project moves to betta status and beyo
nd. This feature (hair-pinning) is an example of a feature that is currently available in the alpha version of ambient and under review for modification as the project evolves.
{{< /warning >}}_

ilrudie commented on content/en/docs/ops/ambient/usage/ztunnel/index.md:
```diff
@@ -0,0 +1,656 @@
+    caption="Ztunnel traffic hairpinning"
+    >}}
+
+### Note on HBONE
```
HBONE section may make more sense as an introductory section. Feels a little unnatural here, after its been used quite a bit with very little explanation.

srampal replied:
Having this earlier was not necessary since all that mattered initially was it was some form of tunneling but otherwise was transparent to the user and didnt need any configuration. Once the user has understood some basics of the ztunnel proxy by the time they get to this point, it helps to add a bit more detail hence imo this location is probably reasonable.

ilrudie reviewed (commented):
Only got through the first 3 sections so far but I wanted to leave the partial review so you can take a look at my comments sooner rather than later. Over all this is much needed content. Thanks for putting it together!

istio-testing commented:
@srampal: The following test **failed**, say `/retest` to rerun all failed tests or `/retest-required` to rerun all mandatory failed tests:

Test name | Commit | Details | Required | Rerun command
--- | --- | --- | --- | ---
lint_istio.io | 886971fb334ac3a93244c2d12be45c983136d88e | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio.io/13635/lint_istio.io/1718025097790361600) | true | `/test lint`

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

srampal commented on content/en/docs/ops/ambient/usage/ztunnel/index.md:
```diff
@@ -0,0 +1,656 @@
+
+## Functional Overview {#functionaloverview}
+
+In this section we provide a functional summary of the control plane and data plane behavior of current version of the ztunnel proxy. Detailed description of architecture internals of the ztunnel proxy is out of scope for this guide.  
```
changed this to 

The functional behavior of the ztunnel proxy can be divided into its data plane behavior and its interaction with the Istio control plane. In this section we take a brief look at these two aspects. Detailed description of internal design of the ztunnel proxy is out of scope for this guide.

srampal commented on content/en/docs/ops/ambient/usage/ztunnel/index.md:
```diff
@@ -0,0 +1,656 @@
+
+1. **K8s CNI restrictions:**  Istio in Ambient mode does not currently work alongside every single Kubernetes CNI plugin implementations Additionally, with some plugins. certain CNI functions (in particular Kubernetes NetworkPolicy and Kubernetes Service Load balancing features) may get transparently bypassed in the presence of Istio Ambient mode. The exact set of supported CNI plugins as well as ...
+ 
+1. **TCP/ IPv4 only:** In the current release, TCP over IPv4 is the only protocol supported for transport on an Istio HBONE tunnel (this includes protocols such as HTTP that run between application layer endpoints on top of the TCP/ IPv4 combination independent of the outermost HTTP used for the HBONE transport between ztunnel proxies).
```
changed this to the following ..

**TCP/ IPv4 only:** In the current release, TCP over IPv4 is the only protocol supported for transport on an Istio secure overlay tunnel (this includes protocols such as HTTP that run between application layer endpoints on top of the TCP/ IPv4 connection).

srampal commented on content/en/docs/ops/ambient/usage/ztunnel/index.md:
```diff
@@ -0,0 +1,656 @@
+
+### Environment used for this guide 
+
+For the examples in this guide, we used a deployment of Istio version 1.19.0 on a `kind` cluster of version 0.20.0 running Kubernetes version 1.27.3. However these should also work on any Kubernetes cluster at version 1.24.0 or later and Istio version 1.18.0 or later.  It would be recommended to have a cluster with more than 1 worker node in order to fully exercise the examples described in this  ...
```
Done

srampal commented on content/en/docs/ops/ambient/usage/ztunnel/index.md:
```diff
@@ -0,0 +1,656 @@
+
+1. **No Istio Multi-cluster support:**  Only single cluster deployments are currently supported for Istio Ambient mode.
+
+1. **K8s CNI restrictions:**  Istio in Ambient mode does not currently work alongside every single Kubernetes CNI plugin implementations Additionally, with some plugins. certain CNI functions (in particular Kubernetes NetworkPolicy and Kubernetes Service Load balancing features) may get transparently bypassed in the presence of Istio Ambient mode. The exact set of supported CNI plugins as well as ...
```
Done

srampal commented on content/en/docs/ops/ambient/usage/ztunnel/index.md:
```diff
@@ -0,0 +1,656 @@
+
+ The ztunnel node proxy is responsible for securely connecting and authenticating workloads within the `ambient` mesh. The ztunnel proxy is written in Rust and is intentionally scoped to handle L3 and L4 functions in the `ambient` mesh such as mTLS, authentication, L4 authorization and telemetry. Ztunnel does not terminate workload HTTP traffic or parse workload HTTP headers. The ztunnel ensures  ...
+
+It is expected that some production use cases of Istio in Ambient mode may be addressed solely via the L4 Secure overlay networking features whereas other use cases will additionally need advanced traffic management and L7 networking features for which the additional waypoint proxies will need to be deployed. This is summarized in the following table. This guide focuses on functionality related t ...
```
Done

srampal commented on content/en/docs/ops/ambient/usage/ztunnel/index.md:
```diff
@@ -0,0 +1,656 @@
+Pods/ workloads using sidecar proxies can co-exist within the same mesh as pods that operate in `ambient` mode.  Mesh pods that use sidecar proxies can also interoperate with pods in the same Istio mesh that are running in ambient mode. The term `ambient` mesh refers to an Istio mesh that has a superset of the capabilities and hence can support mesh pods that use either type of proxying.  
+{{< /tip >}}
+
+ The ztunnel node proxy is responsible for securely connecting and authenticating workloads within the `ambient` mesh. The ztunnel proxy is written in Rust and is intentionally scoped to handle L3 and L4 functions in the `ambient` mesh such as mTLS, authentication, L4 authorization and telemetry. Ztunnel does not terminate workload HTTP traffic or parse workload HTTP headers. The ztunnel ensures  ...
```
Fixed

srampal commented on content/en/docs/ops/ambient/usage/ztunnel/index.md:
```diff
@@ -0,0 +1,656 @@
+
+### Environment used for this guide 
+
+For the examples in this guide, we used a deployment of Istio version 1.19.0 on a `kind` cluster of version 0.20.0 running Kubernetes version 1.27.3. However these should also work on any Kubernetes cluster at version 1.24.0 or later and Istio version 1.18.0 or later.  It would be recommended to have a cluster with more than 1 worker node in order to fully exercise the examples described in this  ...
```
Changed this to ...

For the examples in this guide, we used a deployment of Istio version 1.19.0 on a `kind` cluster of version 0.20.0 running Kubernetes version 1.27.3. The minimum Istio version needed for ambient functions is 1.18.0 and the minimum Kubernetes version needed is 1.24.0.  It would be recommended to have a cluster with more than 1 worker node in order to fully exercise the examples described in this guide. Refer to the [installation user guide](/docs/ops/ambient/usage/install/) or [getting started guide](/docs/ops/ambient/getting-started/) for information on installing Istio in ambient mode on a Kubernetes cluster.

srampal commented on content/en/docs/ops/ambient/usage/ztunnel/index.md:
```diff
@@ -0,0 +1,656 @@
+    >}}
+
+
+Each instance of the ztunnel proxy uses the Envoy xDS apis to receive certificates, discovery and configuration information from the istio control plane (`istiod`) on behalf of all pods and endpoints associated with it. In particular the ztunnel proxy obtains mTLS certificates for all Service accounts of all pods that are associated with it. Since a single ztunnel proxy performs both the data pla ...
```
> I think it's a bit confusing to say that ztunnel performs "control plane operations". I think you mean that it handles the tasks of communicating with the control plane.

changed it to ..
Since a single ztunnel proxy is shared by pods that use multiple different service accounts on a node, it is a multi-tenant component of the mesh infrastructure in contrast with Istio side car proxies that handle control plane and data plane operations on a per application endpoint or pod basis.

srampal replied:
Okay, used the suggested text.

srampal commented on content/en/docs/ops/ambient/usage/ztunnel/index.md:
```diff
@@ -0,0 +1,656 @@
+
+Each instance of the ztunnel proxy uses the Envoy xDS apis to receive certificates, discovery and configuration information from the istio control plane (`istiod`) on behalf of all pods and endpoints associated with it. In particular the ztunnel proxy obtains mTLS certificates for all Service accounts of all pods that are associated with it. Since a single ztunnel proxy performs both the data pla ...
+
+It is also worth noting that in Ambient mode, a simplified set of resources are used in the xDS apis for ztunnel proxy configuration. This results in improved performance (having to transmit and process a much smaller set of information that is sent from `istiod` to the ztunnel proxies) and improved troulbeshooting. 
```
Okay, done.

srampal commented on content/en/docs/ops/ambient/usage/ztunnel/index.md:
```diff
@@ -0,0 +1,656 @@
+    caption="Basic ztunnel L4-only datapath"
+    >}}
+
+The figure depicts ambient pod workloads running on two nodes W1 and W2 of a Kubernetes cluster. There is a single instance of the ztunnel proxy on each node. In this scenario, application client pods C1, C2 and C3 need to access a service provided by pod S1 and there is no requirement for advanced L7 services such as L7 traffic routing or L7 traffic manaagementand hence no Waypoint proxy needed  ...
```
Okay .. done.

srampal commented on content/en/docs/ops/ambient/usage/ztunnel/index.md:
```diff
@@ -0,0 +1,656 @@
+
+The figure depicts ambient pod workloads running on two nodes W1 and W2 of a Kubernetes cluster. There is a single instance of the ztunnel proxy on each node. In this scenario, application client pods C1, C2 and C3 need to access a service provided by pod S1 and there is no requirement for advanced L7 services such as L7 traffic routing or L7 traffic manaagementand hence no Waypoint proxy needed  ...
+
+The figure shows that pods C1 and C2 running on node W1 connect with pod S1 running on node W2 and their TCP traffic is tunneled through a single shared HBONE tunnel instance that has been created between the ztunnel proxy pods of each node. Mutual TLS (mTLS) is used for encryption as well as mutual authentication of traffic getting tunneled. SPIFFE identities are used as the workload identities  ...
```
Okay ,, done

istio-testing commented:
PR needs rebase.

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository.
</details>

istio-testing commented:
@srampal: The following tests **failed**, say `/retest` to rerun all failed tests or `/retest-required` to rerun all mandatory failed tests:

Test name | Commit | Details | Required | Rerun command
--- | --- | --- | --- | ---
lint_istio.io | 886971fb334ac3a93244c2d12be45c983136d88e | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio.io/13635/lint_istio.io/1718025097790361600) | true | `/test lint`
update-ref-docs-dry-run_istio.io | 4e1a8fe54cb9d9967e285b2fbf49683a520fa4f5 | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio.io/13635/update-ref-docs-dry-run_istio.io/1719438264982573056) | false | `/test update-ref-docs-dry-run`

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

istio-testing commented:
@srampal: The following tests **failed**, say `/retest` to rerun all failed tests or `/retest-required` to rerun all mandatory failed tests:

Test name | Commit | Details | Required | Rerun command
--- | --- | --- | --- | ---
update-ref-docs-dry-run_istio.io | 4e1a8fe54cb9d9967e285b2fbf49683a520fa4f5 | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio.io/13635/update-ref-docs-dry-run_istio.io/1719438264982573056) | false | `/test update-ref-docs-dry-run`
lint_istio.io | 4e1a8fe54cb9d9967e285b2fbf49683a520fa4f5 | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio.io/13635/lint_istio.io/1719438264898686976) | true | `/test lint`

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

istio-testing commented:
@srampal: The following tests **failed**, say `/retest` to rerun all failed tests or `/retest-required` to rerun all mandatory failed tests:

Test name | Commit | Details | Required | Rerun command
--- | --- | --- | --- | ---
update-ref-docs-dry-run_istio.io | 4e1a8fe54cb9d9967e285b2fbf49683a520fa4f5 | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio.io/13635/update-ref-docs-dry-run_istio.io/1719438264982573056) | false | `/test update-ref-docs-dry-run`
lint_istio.io | 4e1a8fe54cb9d9967e285b2fbf49683a520fa4f5 | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio.io/13635/lint_istio.io/1719438264898686976) | true | `/test lint`
gencheck_istio.io | 4e1a8fe54cb9d9967e285b2fbf49683a520fa4f5 | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio.io/13635/gencheck_istio.io/1719438264940630016) | true | `/test gencheck`

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

istio-testing commented:
@srampal: The following tests **failed**, say `/retest` to rerun all failed tests or `/retest-required` to rerun all mandatory failed tests:

Test name | Commit | Details | Required | Rerun command
--- | --- | --- | --- | ---
update-ref-docs-dry-run_istio.io | 4e1a8fe54cb9d9967e285b2fbf49683a520fa4f5 | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio.io/13635/update-ref-docs-dry-run_istio.io/1719438264982573056) | false | `/test update-ref-docs-dry-run`
lint_istio.io | 4e1a8fe54cb9d9967e285b2fbf49683a520fa4f5 | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio.io/13635/lint_istio.io/1719438264898686976) | true | `/test lint`
gencheck_istio.io | 86601318569ef19b5addc9918ad2dd175158d35b | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio.io/13635/gencheck_istio.io/1719441613853298688) | true | `/test gencheck`

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

istio-testing commented:
@srampal: The following tests **failed**, say `/retest` to rerun all failed tests or `/retest-required` to rerun all mandatory failed tests:

Test name | Commit | Details | Required | Rerun command
--- | --- | --- | --- | ---
lint_istio.io | 4e1a8fe54cb9d9967e285b2fbf49683a520fa4f5 | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio.io/13635/lint_istio.io/1719438264898686976) | true | `/test lint`
gencheck_istio.io | 86601318569ef19b5addc9918ad2dd175158d35b | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio.io/13635/gencheck_istio.io/1719441613853298688) | true | `/test gencheck`
update-ref-docs-dry-run_istio.io | 86601318569ef19b5addc9918ad2dd175158d35b | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio.io/13635/update-ref-docs-dry-run_istio.io/1719441613899436032) | false | `/test update-ref-docs-dry-run`

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

istio-testing commented:
@srampal: The following tests **failed**, say `/retest` to rerun all failed tests or `/retest-required` to rerun all mandatory failed tests:

Test name | Commit | Details | Required | Rerun command
--- | --- | --- | --- | ---
gencheck_istio.io | 86601318569ef19b5addc9918ad2dd175158d35b | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio.io/13635/gencheck_istio.io/1719441613853298688) | true | `/test gencheck`
update-ref-docs-dry-run_istio.io | 86601318569ef19b5addc9918ad2dd175158d35b | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio.io/13635/update-ref-docs-dry-run_istio.io/1719441613899436032) | false | `/test update-ref-docs-dry-run`
lint_istio.io | 86601318569ef19b5addc9918ad2dd175158d35b | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio.io/13635/lint_istio.io/1719441613802967040) | true | `/test lint`

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

istio-testing commented:
@srampal: The following test **failed**, say `/retest` to rerun all failed tests or `/retest-required` to rerun all mandatory failed tests:

Test name | Commit | Details | Required | Rerun command
--- | --- | --- | --- | ---
lint_istio.io | 486fb710301a23199facc76cd1fe1bc4eb2d4fdf | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio.io/13635/lint_istio.io/1719523219548934144) | true | `/test lint`

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

bleggett commented on content/en/docs/ops/ambient/usage/ztunnel/index.md:
```diff
@@ -0,0 +1,675 @@
+sleep-69cfb4968f-rhhhp     1/1     Running   0          78m
+{{< /text >}}
+
+Further, we see that after this, we still see only 1 container per application pod and the uptime of these pods indicates these were not restarted in order to enable `ambient` mode (unlike `sidecar` mode which does restart application pods when the sidecar proxies are injected). This results in better user experience and operational performance since `ambient` mode can seamlessly be enabled (or d ...
```
```suggestion
Note that after `ambient` is enabled for the namespace, every application pod still only has 1 container, and the uptime of these pods indicates these were not restarted in order to enable `ambient` mode (unlike `sidecar` mode which does restart application pods when the sidecar proxies are injected). This results in better user experience and operational performance since `ambient` mode can seamlessly be enabled (or disabled) completely transparently as far as the application pods are concerned. 

Initiate a `curl` request again to from a client pod to to a service endpoint to verify traffic continues to flow while in `ambient` mode:
```

srampal replied:
made the suggested changes

bleggett commented on content/en/docs/ops/ambient/usage/ztunnel/index.md:
```diff
@@ -0,0 +1,675 @@
+
+The ztunnel node proxy is responsible for securely connecting and authenticating workloads within the `ambient` mesh. The ztunnel proxy is written in Rust and is intentionally scoped to handle L3 and L4 functions in the `ambient` mesh such as mTLS, authentication, L4 authorization and telemetry. Ztunnel does not terminate workload HTTP traffic or parse workload HTTP headers. The ztunnel ensures L ...
+
+We anticipate that some use cases of Istio in Ambient mode may be addressed solely via the L4 secure overlay networking features whereas other use cases requiring advanced traffic management and L7 networking features will necessitate deployment of a waypoint proxy. This guide focuses on functionality related to the L4 secure overlay network using ztunnel proxies. In this guide we refer to L7 onl ...
```
```suggestion
Some use cases of Istio in Ambient mode may be addressed solely via the L4 secure overlay networking features and will not need L7 features, and will not require deployment of a waypoint proxy. Other use cases requiring advanced traffic management and L7 networking features will require deployment of a waypoint proxy. This guide focuses on functionality related to the L4 secure overlay network using ztunnel proxies. In this guide we refer to L7 only when needed to describe some L4 ztunnel function. Other guides are dedicated to cover the advanced L7 networking functions and the use of waypoint proxies.
```

srampal replied:
made the suggested changes

bleggett commented on content/en/docs/ops/ambient/usage/ztunnel/index.md:
```diff
@@ -0,0 +1,675 @@
+
+### Environment used for this guide
+
+For the examples in this guide, we used a deployment of Istio version 1.19.0 on a `kind` cluster of version 0.20.0 running Kubernetes version 1.27.3. The minimum Istio version needed for ambient functions is 1.18.0 and the minimum Kubernetes version needed is 1.24.0. It would be recommended to have a cluster with more than 1 worker node in order to fully exercise the examples described in this gu ...
```
```suggestion
The examples in this guide used a deployment of Istio version 1.19.0 on a `kind` cluster of version 0.20.0 running Kubernetes version 1.27.3. 

The minimum Istio version needed for ambient functions is `1.18.0` and the minimum Kubernetes version needed is `1.24.0`. The examples below require a cluster with more than 1 worker node in order to explain how cross-node traffic operates. Refer to the [installation user guide](/docs/ops/ambient/usage/install/) or [getting started guide](/docs/ops/ambient/getting-started/) for information on installing Istio in ambient mode on a Kubernetes cluster.
```

srampal replied:
made the suggested changes

bleggett commented on content/en/docs/ops/ambient/usage/ztunnel/index.md:
```diff
@@ -0,0 +1,675 @@
+
+## Functional Overview {#functionaloverview}
+
+The functional behavior of the ztunnel proxy can be divided into its data plane behavior and its interaction with the Istio control plane. In this section we take a brief look at these two aspects. Detailed description of internal design of the ztunnel proxy is out of scope for this guide.
```
```suggestion
The functional behavior of the ztunnel proxy can be divided into its data plane behavior and its interaction with the Istio control plane. This section takes a brief look at these two aspects - detailed description of internal design of the ztunnel proxy is out of scope for this guide.
```

srampal replied:
made the suggested changes

dhawton commented on content/en/docs/ops/ambient/usage/ztunnel/index.md:
```diff
@@ -0,0 +1,675 @@
+{{< warning >}}
+Ambient is currently in [alpha status](/docs/releases/feature-stages/#feature-phase-definitions).
+
+Please **do not run ambient in production** and be sure to thoroughly review the [feature phase definitions](/docs/releases/feature-stages/#feature-phase-definitions) before use. In particular, there are known performance, stability, and security issues in the `alpha` release. There are also functional caveats some of which are listed in the [Caveats section](#caveats) of this guide. There are al ...
```
In addition, you don't need to backtick ambient.

srampal replied:
Okay, made the suggested change and removed most of the backticks for ambient.

srampal commented on content/en/docs/ops/ambient/usage/ztunnel/index.md:
```diff
@@ -0,0 +1,675 @@
+
+## Introduction {#introsection}
+
+This guide describes the functionality and usage of the ztunnel proxy and Layer-4 networking functions using Istio Ambient mesh. We use a sample user journey to describe these functions hence it would be useful to go through this guide in sequence. However we provide links to the sections below in case the reader would like to jump to the appropriate section.
```
Made the suggested changes

srampal commented on content/en/docs/ops/ambient/usage/ztunnel/index.md:
```diff
@@ -0,0 +1,675 @@
+Pods/ workloads using sidecar proxies can co-exist within the same mesh as pods that operate in `ambient` mode. Mesh pods that use sidecar proxies can also interoperate with pods in the same Istio mesh that are running in ambient mode. The term `ambient` mesh refers to an Istio mesh that has a superset of the capabilities and hence can support mesh pods that use either type of proxying.
+{{< /tip >}}
+
+The ztunnel node proxy is responsible for securely connecting and authenticating workloads within the `ambient` mesh. The ztunnel proxy is written in Rust and is intentionally scoped to handle L3 and L4 functions in the `ambient` mesh such as mTLS, authentication, L4 authorization and telemetry. Ztunnel does not terminate workload HTTP traffic or parse workload HTTP headers. The ztunnel ensures L ...
```
made the suggested changes

srampal commented on content/en/docs/ops/ambient/usage/ztunnel/index.md:
```diff
@@ -0,0 +1,675 @@
+
+The figure shows that pods C1 and C2 running on node W1 connect with pod S1 running on node W2 and their TCP traffic is tunneled through a single shared HBONE tunnel instance that has been created between the ztunnel proxy pods of each node. Mutual TLS (mTLS) is used for encryption as well as mutual authentication of traffic being tunneled. SPIFFE identities are used to identify the workloads on  ...
+
+We note that the figure also shows traffic from pod C3 destined to destination pod S1 on worker node W2 also traverses the local ztunnel proxy instance so that L4 traffic management functions such as L4 Authorization and L4 Telemetry are enforced identically on traffic whether or not it crosses a node boundary.
```
made the suggested changes

srampal commented on content/en/docs/ops/ambient/usage/ztunnel/index.md:
```diff
@@ -0,0 +1,675 @@
+
+#### Ztunnel datapath hair-pinning
+
+Finally we illustrate the concept of hair=pinning in the data plane via the next figure. We noted earlier that traffic is always sent to a destination pod by first sending it to the ztunnel proxy on the same node as the destination pod. But what if the sender is either completely outside the Istio ambient mesh and hence does not initiate HBONE tunnels to the destination ztunnel first ? or what if ...
```
made the suggested changes

srampal commented on content/en/docs/ops/ambient/usage/ztunnel/index.md:
```diff
@@ -0,0 +1,675 @@
+
+### Basic application deployment without Ambient
+
+Let's first deploy a simple HTTP client server application without making it part of the Istio ambient mesh. We can pick from the apps in the samples folder of the Istio repository. Execute the following examples from the top of a local Istio repository or Istio folder created by downloading the istioctl client as described in Istio guides.
```
Made the suggested change

srampal commented on content/en/docs/ops/ambient/usage/ztunnel/index.md:
```diff
@@ -0,0 +1,675 @@
+$ kubectl scale deployment sleep --replicas=2 -n ambient-demo
+{{< /text >}}
+
+These manifests should deploy the `sleep` and `notsleep` pods which we shall use as clients for the httpbin service pod (for simplicity, the cli outputs have been deleted in the code samples above). We also create multiple replicas of the client deployment in order to exercise various scenarios.
```
made the suggested change

srampal commented on content/en/docs/ops/ambient/usage/ztunnel/index.md:
```diff
@@ -0,0 +1,675 @@
+
+### Enabling Ambient for an application
+
+We now enable `ambient` for the application deployed in the prior subsection by simply adding the label `istio.io/dataplane-mode=ambient` to the application's namespace as shown below. A namespace should not be enabled for both sidecar mode and ambient mode at the same time. For now, we focus on using `ambient` mode only. In a following subsection we describe how conflicts are resolved in hybrid  ...
```
made the suggested changes

istio-testing commented:
@srampal: The following test **failed**, say `/retest` to rerun all failed tests or `/retest-required` to rerun all mandatory failed tests:

Test name | Commit | Details | Required | Rerun command
--- | --- | --- | --- | ---
lint_istio.io | 09c55dfde7dfb32983b76e49cb047bfd4a5a5247 | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio.io/13635/lint_istio.io/1721752167942983680) | true | `/test lint`

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

dhawton commented on content/en/docs/ops/ambient/usage/ztunnel/index.md:
```diff
@@ -0,0 +1,620 @@
+
+Again, we can further verify from viewing the logs of the ztunnel pod (not shown in the example) at the destination node that traffic does in fact use the HBONE and CONNECT based path from the sidecar proxy based source client pod to the ambient based destination service pod. Additionally not shown but we can also verify that unlike the previous subsection, in this case even if we apply a PeerAut ...
+
+### North-South Ingress Gateway to Ambient backend pods {#nsingress2ambient}
```
```suggestion
### North-South Ingress Gateway to ambient backend pods {#nsingress2ambient}
```

srampal replied:
Done

dhawton commented on content/en/docs/ops/ambient/usage/ztunnel/index.md:
```diff
@@ -0,0 +1,620 @@
+
+We now describe a use case for North-South traffic with an Istio Gateway exposing the httpbin service via the Kubernetes Gateway API. The gateway itself is running in a non-Ambient namespace and may be an existing gateway that is also exposing other services that are provided by non-ambient pods. Hence this example shows that ambient workloads can also interoperate with Istio gateways that need n ...
+
+For this example, we will use `metallb` to provide a load balancer service on an IP addresses that is reachable from outside the cluster. However the same example also works with other forms of North-South load balancing options. For the example below, we assume that you have already installed `metallb` in this cluster to provided the load balancer service including a pool of IP addresses for `me ...
```
```suggestion
For this example, we will use `metallb` to provide a load balancer service on an IP addresses that is reachable from outside the cluster. However the same example also works with other forms of North-South load balancing options. For the example below, we assume that you have already installed `metallb` in this cluster to provided the load balancer service including a pool of IP addresses for `metallb` to use for exposing services externally. Refer to the [`metallb` guide for kind](https://kind.sigs.k8s.io/docs/user/loadbalancer/) for instructions on setting up `metallb` on kind clusters or refer to the instructions from the [`metallb` documentation](https://metallb.universe.tf/installation/) appropriate for your environment.
```

srampal replied:
Done

dhawton commented on content/en/docs/ops/ambient/usage/ztunnel/index.md:
```diff
@@ -0,0 +1,620 @@
+
+For this example, we will use `metallb` to provide a load balancer service on an IP addresses that is reachable from outside the cluster. However the same example also works with other forms of North-South load balancing options. For the example below, we assume that you have already installed `metallb` in this cluster to provided the load balancer service including a pool of IP addresses for `me ...
+
+In this example we use the Kubernetes Gateway API for configuring the N-S gateway. Since this API is not currently provided as default in Kubernetes and Kind distributions, we have to install the API CRDs first as shown in the example.
```
```suggestion
In this example we use the Kubernetes Gateway API for configuring the N-S gateway. Since this API is not currently provided as default in Kubernetes and kind distributions, we have to install the API CRDs first as shown in the example.
```

srampal replied:
Done

dhawton commented on content/en/docs/ops/ambient/usage/_index.md:
```diff
@@ -1,9 +1,7 @@
 owner: istio/wg-networking-maintainers
 test: n/a
 ---
-
-This page is under construction.
```
I don't really like empty pages.. can you put this back in and add these to the other shells you created?

srampal replied:
Done

dhawton commented on content/en/docs/ops/ambient/usage/install/index.md:
```diff
@@ -0,0 +1,8 @@
+owner: istio/wg-networking-maintainers
+test: no
+---
+
```
Add something here

srampal replied:
Done

dhawton commented on content/en/docs/ops/ambient/usage/waypoint/index.md:
```diff
@@ -0,0 +1,8 @@
+owner: istio/wg-networking-maintainers
+test: no
+---
+
```
Add something here as well

srampal replied:
Done

srampal commented on content/en/docs/ops/ambient/usage/ztunnel/index.md:
```diff
@@ -0,0 +1,620 @@
+
+## Introduction {#introsection}
+
+This guide describes in-depth the functionality and usage of the ztunnel proxy and Layer-4 networking functions in Istio `Ambient` mesh. To simply try out Istio Ambient mesh, follow the [Ambient Quickstart](/docs/ops/ambient/getting-started/) instead. This guide follows a user journey and works through multiple examples to detail the design and architecture of Istio Ambient. It is highly recommen ...
```
Done

srampal commented on content/en/docs/ops/ambient/usage/ztunnel/index.md:
```diff
@@ -0,0 +1,620 @@
+The ztunnel (Zero Trust Tunnel) component is a purpose-built per-node proxy for Istio ambient mesh. Since workload pods no longer require proxies running in sidecars in order to participate in the mesh, Istio in ambient mode is informally also referred to as "sidecar-less" mesh.
+
+{{< tip >}}
+Pods/ workloads using sidecar proxies can co-exist within the same mesh as pods that operate in ambient mode. Mesh pods that use sidecar proxies can also interoperate with pods in the same Istio mesh that are running in ambient mode. The term ambient mesh refers to an Istio mesh that has a superset of the capabilities and hence can support mesh pods that use either type of proxying.
```
Done

srampal commented on content/en/docs/ops/ambient/usage/ztunnel/index.md:
```diff
@@ -0,0 +1,620 @@
+
+The ztunnel node proxy is responsible for securely connecting and authenticating workloads within the ambient mesh. The ztunnel proxy is written in Rust and is intentionally scoped to handle L3 and L4 functions in the ambient mesh such as mTLS, authentication, L4 authorization and telemetry. Ztunnel does not terminate workload HTTP traffic or parse workload HTTP headers. The ztunnel ensures L3 an ...
+
+Some use cases of Istio in Ambient mode may be addressed solely via the L4 secure overlay networking features, and will not need L7 features thereby not requiring deployment of a waypoint proxy. Other use cases requiring advanced traffic management and L7 networking features will require deployment of a waypoint proxy. This guide focuses on functionality related to the L4 secure overlay network u ...
```
Done

srampal commented on content/en/docs/ops/ambient/usage/ztunnel/index.md:
```diff
@@ -0,0 +1,620 @@
+
+## Current Caveats {#caveats}
+
+Ztunnel proxies are automatically installed when one of the supported installation methods is used to install Istio ambient mesh. The minimum Istio version required for Istio ambient mode is `1.18.0`. In general Istio in ambient mode supports the existing Istio APIs that are supported in sidecar proxy mode. Since the Ambient functionality is currently at an alpha release level, the following is a ...
```
Done

srampal commented on content/en/docs/ops/ambient/usage/ztunnel/index.md:
```diff
@@ -0,0 +1,620 @@
+
+1. **Kubernetes (K8s) only:** Istio in ambient mode is currently only supported for deployment on Kubernetes clusters. Deployment on non-Kubernetes endpoints such as virtual machines is not currently supported.
+
+1. **No Istio Multi-cluster support:** Only single cluster deployments are currently supported for Istio Ambient mode.
```
Done

srampal commented on content/en/docs/ops/ambient/usage/ztunnel/index.md:
```diff
@@ -0,0 +1,620 @@
+
+1. **No Istio Multi-cluster support:** Only single cluster deployments are currently supported for Istio Ambient mode.
+
+1. **K8s CNI restrictions:** Istio in Ambient mode does not currently work with every Kubernetes CNI implementation. Additionally, with some plugins, certain CNI functions (in particular Kubernetes NetworkPolicy and Kubernetes Service Load balancing features) may get transparently bypassed in the presence of Istio Ambient mode. The exact set of supported CNI plugins as well as any CNI feature cav ...
```
Done

srampal commented on content/en/docs/ops/ambient/usage/ztunnel/index.md:
```diff
@@ -0,0 +1,620 @@
+
+1. **K8s CNI restrictions:** Istio in Ambient mode does not currently work with every Kubernetes CNI implementation. Additionally, with some plugins, certain CNI functions (in particular Kubernetes NetworkPolicy and Kubernetes Service Load balancing features) may get transparently bypassed in the presence of Istio Ambient mode. The exact set of supported CNI plugins as well as any CNI feature cav ...
+
+1. **TCP/ IPv4 only:** In the current release, TCP over IPv4 is the only protocol supported for transport on an Istio secure overlay tunnel (this includes protocols such as HTTP that run between application layer endpoints on top of the TCP/ IPv4 connection).
```
Done

srampal commented on content/en/docs/ops/ambient/usage/ztunnel/index.md:
```diff
@@ -0,0 +1,620 @@
+
+1. **TCP/ IPv4 only:** In the current release, TCP over IPv4 is the only protocol supported for transport on an Istio secure overlay tunnel (this includes protocols such as HTTP that run between application layer endpoints on top of the TCP/ IPv4 connection).
+
+1. **No dynamic switching to ambient mode:** Ambient mode can only be enabled on a new Istio mesh control plane that is deployed using ambient profile or ambient helm configuration. An existing Istio mesh deployed using a pre-ambient profile for instance can not be dynamically switched to also enable ambient mode operation.
```
Done

srampal commented on content/en/docs/ops/ambient/usage/ztunnel/index.md:
```diff
@@ -0,0 +1,620 @@
+
+1. **No dynamic switching to ambient mode:** Ambient mode can only be enabled on a new Istio mesh control plane that is deployed using ambient profile or ambient helm configuration. An existing Istio mesh deployed using a pre-ambient profile for instance can not be dynamically switched to also enable ambient mode operation.
+
+1. **Restrictions with Istio PeerAuthentication:** as of the time of writing, the PeerAuthentication resource is not supported by all components (i.e. Waypoint proxies) in Istio ambient mode. Hence it is recommended to only use the STRICT mTLS mode currently. Like many of the other alpha stage caveats, this shall be addressed as the feature moves toward beta status.
```
Done

srampal commented on content/en/docs/ops/ambient/usage/ztunnel/index.md:
```diff
@@ -0,0 +1,620 @@
+
+1. **Restrictions with Istio PeerAuthentication:** as of the time of writing, the PeerAuthentication resource is not supported by all components (i.e. Waypoint proxies) in Istio ambient mode. Hence it is recommended to only use the STRICT mTLS mode currently. Like many of the other alpha stage caveats, this shall be addressed as the feature moves toward beta status.
+
+1. **Istioctl CLI gaps:** There may be some minor functional gaps in areas such as Istio CLI output displays when it comes to displaying or monitoring Istio's ambient mode related information. These will be addressed as the feature matures.
```
Done

srampal commented on content/en/docs/ops/ambient/usage/ztunnel/index.md:
```diff
@@ -0,0 +1,620 @@
+
+The ztunnel proxy uses xDS APIs to communicate with the Istio control plane (`istiod`). This enables the fast, dynamic configuration updates required in modern distributed systems. The ztunnel proxy also obtains mTLS certificates for the Service Accounts of all pods that are scheduled on its Kubernetes node using xDS. A single ztunnel proxy may implement L4 data plane functionality on behalf of a ...
+
+It is also worth noting that in Ambient mode, a simplified set of resources are used in the xDS APIs for ztunnel proxy configuration. This results in improved performance (having to transmit and process a much smaller set of information that is sent from istiod to the ztunnel proxies) and improved troubleshooting.
```
Done

srampal commented on content/en/docs/ops/ambient/usage/ztunnel/index.md:
```diff
@@ -0,0 +1,620 @@
+
+The figure depicts ambient pod workloads running on two nodes W1 and W2 of a Kubernetes cluster. There is a single instance of the ztunnel proxy on each node. In this scenario, application client pods C1, C2 and C3 need to access a service provided by pod S1 and there is no requirement for advanced L7 features such as L7 traffic routing or L7 traffic management so no Waypoint proxy is needed.
+
+The figure shows that pods C1 and C2 running on node W1 connect with pod S1 running on node W2 and their TCP traffic is tunneled through a single shared HBONE tunnel instance that has been created between the ztunnel proxy pods of each node. Mutual TLS (mTLS) is used for encryption as well as mutual authentication of traffic being tunneled. SPIFFE identities are used to identify the workloads on  ...
```
Done

srampal commented on content/en/docs/ops/ambient/usage/ztunnel/index.md:
```diff
@@ -0,0 +1,620 @@
+
+### Note on HBONE {#hbonesection}
+
+HBONE (HTTP Based Overlay Network Encapsulation) is an Istio and Ambient specific term. It refers to the use of standard HTTP tunneling via the [HTTP CONNECT](https://developer.mozilla.org/en-US/docs/Web/HTTP/Methods/CONNECT) method to transparently tunnel application packets/ byte streams. In its current implementation within Istio, it transports TCP packets only by tunneling these transparently ...
```
Done

srampal commented on content/en/docs/ops/ambient/usage/ztunnel/index.md:
```diff
@@ -0,0 +1,620 @@
+caption="HBONE L3 packet format"
+>}}
+
+In future Istio Ambient may also support [HTTP/3 (QUIC)](https://datatracker.ietf.org/doc/html/rfc9114) based transport and will be used to transport all types of L3 and L4 packets including native IPv4, IPv6, UDP by leveraging new standards such as CONNECT-UDP and CONNECT-IP being developed as part of the [IETF MASQUE](https://ietf-wg-masque.github.io/) working group. Such additional use cases o ...
```
This text was mentioned on the slack channel as the proposed text to go into the guide and John Howard and others said yes, so I have left it in there. I could take it out or just leave something like the last line if you feel strongly about it.

srampal commented on content/en/docs/ops/ambient/usage/ztunnel/index.md:
```diff
@@ -0,0 +1,620 @@
+    <title>httpbin.org</title>
+{{< /text >}}
+
+### Enabling Ambient for an application
```
Done

srampal commented on content/en/docs/ops/ambient/usage/ztunnel/index.md:
```diff
@@ -0,0 +1,620 @@
+
+### Viewing ztunnel proxy state
+
+As indicated previously, the ztunnel proxy on each node gets configuration and discovery information from the istiod component via `xDS` APIs. Use the "istioctl proxy-config" command shown below to view discovered workloads as seen by a ztunnel proxy as well as secrets holding the TLS certificates that the ztunnel proxy has received from the istiod control plane to use in mTLS signaling on behalf ...
```
Done

srampal commented on content/en/docs/ops/ambient/usage/ztunnel/index.md:
```diff
@@ -0,0 +1,620 @@
+spiffe://cluster.local/ns/istio-system/sa/ztunnel     Cert Chain     Available     true           ec30e0e1b7105e3dce4425b5255287c6     2033-09-16T18:26:19Z     2023-09-19T18:26:19Z
+{{< /text >}}
+
+Using these cli commands, a user can check that ztunnel proxies are getting configured with all the expected workloads and TLS certificates and missing information can be used for troubleshooting to explain any potential observed networking errors. A user may also use the `all` option to view all parts of the proxy-config with a single cli command and the json output formatter as shown in the exa ...
```
Done

srampal commented on content/en/docs/ops/ambient/usage/ztunnel/index.md:
```diff
@@ -0,0 +1,620 @@
+$ istioctl proxy-config all "$ZTUNNEL".istio-system -o json | jq
+{{< /text >}}
+
+Note that when used with a ztunnel proxy instance, not all options of the "istioctl proxy-config" cli are supported since some apply only to sidecar proxies.
```
Done

srampal commented on content/en/docs/ops/ambient/usage/ztunnel/index.md:
```diff
@@ -0,0 +1,620 @@
+
+### Viewing Istiod state for ztunnel xDS resources
+
+Sometimes an advanced user may want to view the state of ztunnel proxy config resources as maintained in the istiod control plane, in the format of the xDS API resources defined specially for ztunnel proxies. This can be done by exec-ing into the istiod pod and obtaining this information from port 15014 for a given ztunnel proxy as shown in the example below. This output can then also be saved an ...
```
Done

srampal commented on content/en/docs/ops/ambient/usage/ztunnel/index.md:
```diff
@@ -0,0 +1,620 @@
+--snip--
+{{< /text >}}
+
+These log messages confirm the traffic indeed used the `ztunnel` proxy in the datapath. Additional fine grained monitoring can be done by checking logs on the specific `ztunnel` proxy instances that are on the same nodes as the source and destination pods of traffic. If these logs are not seen, then a possibility is that traffic redirection may not be working correctly. Detailed description of mo ...
```
Done

srampal commented on content/en/docs/ops/ambient/usage/ztunnel/index.md:
```diff
@@ -0,0 +1,620 @@
+### Monitoring and Telemetry via Prometheus, Grafana, Kiali
+
+In addition to checking ztunnel logs and other monitoring options noted above, one can also use normal Istio monitoring and telemetry functions to monitor application traffic within an Istio Ambient mesh.
+The use of Istio in Ambient mode does not change this behavior. Since this functionality is largely unchanged in Istio ambient mode from Istio sidecar mode , we do not repeat these details in this guide. Please refer to [Prometheus](/docs/ops/integrations/prometheus/#installation) and [Kiali](/docs/ops/integrations/kiali/#installation) for information on installation of Prometheus and Kiali servi ...
```
Done

srampal commented on content/en/docs/ops/ambient/usage/ztunnel/index.md:
```diff
@@ -0,0 +1,620 @@
+In addition to checking ztunnel logs and other monitoring options noted above, one can also use normal Istio monitoring and telemetry functions to monitor application traffic within an Istio Ambient mesh.
+The use of Istio in Ambient mode does not change this behavior. Since this functionality is largely unchanged in Istio ambient mode from Istio sidecar mode , we do not repeat these details in this guide. Please refer to [Prometheus](/docs/ops/integrations/prometheus/#installation) and [Kiali](/docs/ops/integrations/kiali/#installation) for information on installation of Prometheus and Kiali servi ...
+
+One point to note is that in case of a service that is only using ztunnel and L4 networking, the Istio metrics reported will currently only be the L4/ TCP metrics (namely istio_tcp_sent_bytes_total, istio_tcp_received_bytes_total, istio_tcp_connections_opened_total, istio_tcp_connections_closed_total). The full set of Istio and Envoy metrics will be reported when a Waypoint proxy or a Gateway pod ...
```
Done

srampal commented on content/en/docs/ops/ambient/usage/ztunnel/index.md:
```diff
@@ -0,0 +1,620 @@
+
+Here we note the logs from the ztunnel proxies first indicating the http CONNECT request to the new destination pod (10.240.1.11) which indicates the setup of the HBONE tunnel to ztunnel on the node hosting the additional destination service pod. This is then followed by logs indicating the client traffic being sent to both 10.240.1.11 and 10.240.2.10 which are the two destination pods providing  ...
+
+This is a Round robin load balancing algorithm and is separate from and independent of any load balancing algorithm that may be configured within a VirtualService's TrafficPolicy field, since as discussed previously, all aspects of VirtualService API objects are instantiated on the Waypoint proxies and not the ztunnel proxies.
```
Done

srampal commented on content/en/docs/ops/ambient/usage/ztunnel/index.md:
```diff
@@ -0,0 +1,620 @@
+
+This is a Round robin load balancing algorithm and is separate from and independent of any load balancing algorithm that may be configured within a VirtualService's TrafficPolicy field, since as discussed previously, all aspects of VirtualService API objects are instantiated on the Waypoint proxies and not the ztunnel proxies.
+
+### Pod selection logic for Ambient and Sidecar modes
```
Done

srampal commented on content/en/docs/ops/ambient/usage/ztunnel/index.md:
```diff
@@ -0,0 +1,620 @@
+1. `ambient` mode is used for a pod if
+- The namespace has label "istio.io/dataplane-mode" == "ambient"
+- The annotation "sidecar.istio.io/status" is not present on the pod
+- "ambient.istio.io/redirection" is not "disabled"
```
Done

srampal commented on content/en/docs/ops/ambient/usage/ztunnel/index.md:
```diff
@@ -0,0 +1,620 @@
+EOF
+{{< /text >}}
+
+The behavior of the AuthorizationPolicy API has the same functional behavior in Istio ambient mode as in sidecar mode. When there is no AuthorizationPolicy provisioned, then the default action is ALLOW. Once the policy above is provisioned, pods matching the selector in the policy (i.e. app:httpbin) only allow traffic explicitly whitelisted which in this case is sources with principal (i.e. ident ...
```
Done

srampal commented on content/en/docs/ops/ambient/usage/ztunnel/index.md:
```diff
@@ -0,0 +1,620 @@
+
+The behavior of the AuthorizationPolicy API has the same functional behavior in Istio ambient mode as in sidecar mode. When there is no AuthorizationPolicy provisioned, then the default action is ALLOW. Once the policy above is provisioned, pods matching the selector in the policy (i.e. app:httpbin) only allow traffic explicitly whitelisted which in this case is sources with principal (i.e. ident ...
+
+Note that this policy performs an explicit ALLOW action on traffic from sources with principal (i.e. identity) of cluster.local/ns/ambient-demo/sa/sleep and hence traffic from all other sources will be denied.
```
Done

srampal commented on content/en/docs/ops/ambient/usage/ztunnel/index.md:
```diff
@@ -0,0 +1,620 @@
+command terminated with exit code 56
+{{< /text >}}
+
+We note that we have not deployed any waypoint proxies and yet this AuthorizationPolicy was getting enforced and this is because this policy only required L4 traffic processing which can be performed by ztunnel proxies. We can further confirm these policy actions by checking ztunnel logs and looking for logs that indicate "RBAC" actions as shown in the following example.
```
Done

srampal commented on content/en/docs/ops/ambient/usage/ztunnel/index.md:
```diff
@@ -0,0 +1,620 @@
+{{< /text >}}
+
+{{< warning >}}
+If an AuthorizationPolicy has been configured that requires any traffic processing beyond L4, and if no waypoint proxies are configured for the destination of the traffic, then ztunnel proxy will simply drop all traffic as a defensive move. Hence check to ensure that either all rules involve L4 processing only or else if non-L4 rules are unavoidable, then waypoint proxies are also configured to h ...
```
Done

srampal commented on content/en/docs/ops/ambient/usage/ztunnel/index.md:
```diff
@@ -0,0 +1,620 @@
+If an AuthorizationPolicy has been configured that requires any traffic processing beyond L4, and if no waypoint proxies are configured for the destination of the traffic, then ztunnel proxy will simply drop all traffic as a defensive move. Hence check to ensure that either all rules involve L4 processing only or else if non-L4 rules are unavoidable, then waypoint proxies are also configured to h ...
+{{< /warning >}}
+
+As an example, if we modify the AuthorizationPolicy to include a check for the HTTP GET method as shown below, then both `sleep` and `notsleep` pods are blocked from sending traffic to the destination httpbin service.
```
Done

srampal commented on content/en/docs/ops/ambient/usage/ztunnel/index.md:
```diff
@@ -0,0 +1,620 @@
+
+We can also confirm via logs of specific ztunnel proxy pods (not shown here) that it is always the ztunnel proxy on the node hosting the destination pod that actually enforces the policy.
+
+Let's go ahead and delete this AuthorizationPolicy before continuing with the rest of the examples in the guide.
```
Done

srampal commented on content/en/docs/ops/ambient/usage/ztunnel/index.md:
```diff
@@ -0,0 +1,620 @@
+$ kubectl delete AuthorizationPolicy allow-sleep-to-httpbin  -n ambient-demo
+{{< /text >}}
+
+## Ambient Interoperability with non-Ambient endpoints {#interop}
```
Done

srampal commented on content/en/docs/ops/ambient/usage/ztunnel/index.md:
```diff
@@ -0,0 +1,620 @@
+
+So far we have looked at use cases where the traffic source and destination pods are both ambient pods. We cover some mixed use cases in this section where ambient endpoints need to communicate with non-ambient endpoints. As with prior examples we will use purely L4 service examples in this section since this guide is primarily focused on ztunnel proxy operation and L4 networking.
+
+1. [East-West non-mesh pod to Ambient mesh pod (and use of PeerAuthentication resource)](#ewnonmesh)
```
Done

srampal commented on content/en/docs/ops/ambient/usage/ztunnel/index.md:
```diff
@@ -0,0 +1,620 @@
+So far we have looked at use cases where the traffic source and destination pods are both ambient pods. We cover some mixed use cases in this section where ambient endpoints need to communicate with non-ambient endpoints. As with prior examples we will use purely L4 service examples in this section since this guide is primarily focused on ztunnel proxy operation and L4 networking.
+
+1. [East-West non-mesh pod to Ambient mesh pod (and use of PeerAuthentication resource)](#ewnonmesh)
+1. [East-West Istio sidecar proxy pod to Ambient mesh pod](#ewside2ambient)
```
Done

srampal commented on content/en/docs/ops/ambient/usage/ztunnel/index.md:
```diff
@@ -0,0 +1,620 @@
+
+1. [East-West non-mesh pod to Ambient mesh pod (and use of PeerAuthentication resource)](#ewnonmesh)
+1. [East-West Istio sidecar proxy pod to Ambient mesh pod](#ewside2ambient)
+1. [North-South Ingress Gateway to Ambient backend pods](#nsingress2ambient)
```
Done

srampal commented on content/en/docs/ops/ambient/usage/ztunnel/index.md:
```diff
@@ -0,0 +1,620 @@
+1. [East-West Istio sidecar proxy pod to Ambient mesh pod](#ewside2ambient)
+1. [North-South Ingress Gateway to Ambient backend pods](#nsingress2ambient)
+
+### East-West non-mesh pod to Ambient mesh pod (and use of PeerAuthentication resource) {#ewnonmesh}
```
Done

srampal commented on content/en/docs/ops/ambient/usage/ztunnel/index.md:
```diff
@@ -0,0 +1,620 @@
+
+### East-West non-mesh pod to Ambient mesh pod (and use of PeerAuthentication resource) {#ewnonmesh}
+
+In the example below, we will access the same httpbin service we have already setup in the prior examples via client `sleep` pods that are running in a separate namespace that is not part of the Istio mesh. This example shows that East-west traffic between Ambient mesh pods and non mesh pods is seamlessly supported. Note that as described previously, this use case leverages the traffic hair-pinni ...
```
Done

srampal commented on content/en/docs/ops/ambient/usage/ztunnel/index.md:
```diff
@@ -0,0 +1,620 @@
+    <title>httpbin.org</title>
+{{< /text >}}
+
+Furthermore, as shown in the example below, we can now add a PeerAuthentication resource with mTLS mode set to strict, in the ambient namespace and confirm that the same client's traffic is now rejected with an error indicating the request was rejected. This is because the client is using simple HTTP to connect to the server instead of an HBONE tunnel with mTLS. This is a possible method that can ...
```
Done

srampal commented on content/en/docs/ops/ambient/usage/ztunnel/index.md:
```diff
@@ -0,0 +1,620 @@
+command terminated with exit code 56
+{{< /text >}}
+
+Change the mTLS mode to PERMISSIVE and confirm that the Ambient pods can once again accept non-mTLS connections including from non-mesh pods in this case.
```
Done

srampal commented on content/en/docs/ops/ambient/usage/ztunnel/index.md:
```diff
@@ -0,0 +1,620 @@
+    <title>httpbin.org</title>
+{{< /text >}}
+
+### East-West Istio sidecar proxy pod to Ambient mesh pod {#ewside2ambient}
```
Done

srampal commented on content/en/docs/ops/ambient/usage/ztunnel/index.md:
```diff
@@ -0,0 +1,620 @@
+We use the same httpbin service from the previous example but now add a client to access this service from another namespace which is labeled for sidecar injection. This also works automatically and transparently as shown in the example below. In this case the sidecar proxy running with the client automatically knows to use the HBONE control plane since the destination has been discovered to be a ...
+
+{{< tip >}}
+For sidecar proxies to use the HBONE/ mTLS signaling option when communicating with ambient destinations, they need to be configured with `ISTIO_META_ENABLE_HBONE` set to true in the proxy metadata. This is automatically set for the user as default in the MeshConfig when using the `ambient` profile, hence the user does not need to do anything additional when using this profile.
```
Done

srampal commented on content/en/docs/ops/ambient/usage/ztunnel/index.md:
```diff
@@ -0,0 +1,620 @@
+    <title>httpbin.org</title>
+{{< /text >}}
+
+Again, we can further verify from viewing the logs of the ztunnel pod (not shown in the example) at the destination node that traffic does in fact use the HBONE and CONNECT based path from the sidecar proxy based source client pod to the ambient based destination service pod. Additionally not shown but we can also verify that unlike the previous subsection, in this case even if we apply a PeerAut ...
```
Done
